pyanaconda/display.py

```diff
@@ -210,41 +207,67 @@ def do_startup_x11_actions():
     # pylint: disable=environment-modify
     os.environ['XDG_CONFIG_DIRS'] = xdg_config_dirs
 
-    def x11_preexec():
+    os.environ["XDG_SESSION_TYPE"] = "wayland"
+
+    def wl_preexec():
         # to set GUI subprocess SIGINT handler
         signal.signal(signal.SIGINT, signal.SIG_IGN)
 
-    childproc = util.startProgram(["gnome-kiosk", "--display", ":1", "--sm-disable", "--x11"],
-                                  env_add={'XDG_DATA_DIRS': xdg_data_dirs},
-                                  preexec_fn=x11_preexec)
-    WatchProcesses.watch_process(childproc, "gnome-kiosk")
+    # lets compile arguments for the run-in-new-session script
+    argv = ["run-in-new-session.py",
+            "--user", "root",
+            "--service", "anaconda",
+            "--session-type", "wayland",
+            "--session-class", "user"]
 
+    if headless:
+        # headless (remote connection) - stay on VT1 where connection info is
+        argv.extend(["--vt", "1"])
+    else:
+        # local display - switch to VT6 & show GUI there
+        argv.extend(["--vt", "6"])
 
-def set_x_resolution(runres):
-    """Set X server screen resolution.
+    # add the generic GNOME Kiosk invocation
+    argv.extend(["gnome-kiosk", "--sm-disable",
+                 "--wayland", "--no-x11",
+                 "--wayland-display", constants.WAYLAND_SOCKET_NAME])
 
-    :param str runres: a resolution specification string
-    """
-    try:
-        log.info("Setting the screen resolution to: %s.", runres)
-        util.execWithRedirect("xrandr", ["-d", ":1", "-s", runres])
-    except RuntimeError:
-        log.error("The X resolution was not set")
-        util.execWithRedirect("xrandr", ["-d", ":1", "-q"])
+    # remote access needs gnome-kiosk to start in headless mode &
+    # configure a virtual monitor
+    if headless:
+        # check virtual monitor  resolution has been set
+        if headless_resolution is None:
+            # use default value
+            headless_resolution = "1280x1024"
+        argv.extend(["--headless", "--virtual-monitor", headless_resolution])
+
+    childproc = util.startProgram(argv, env_add={'XDG_DATA_DIRS': xdg_data_dirs},
+                                  preexec_fn=wl_preexec)
+    WatchProcesses.watch_process(childproc, argv[0])
+
+    for _i in range(0, int(timeout / 0.1)):
+        wl_socket_path = os.path.join(os.getenv("XDG_RUNTIME_DIR"), constants.WAYLAND_SOCKET_NAME)
+        if os.path.exists(wl_socket_path):
+            return
 
+        time.sleep(0.1)
 
-def do_extra_x11_actions(runres, gui_mode):
-    """Perform X11 actions not related to startup.
+    WatchProcesses.unwatch_process(childproc)
+    childproc.terminate()
+    raise TimeoutError("Timeout trying to start gnome-kiosk")
+
+
+def set_resolution(runres):
```

Comment:
This doesn't look necessary, since it should be handled via the RDP protocol.

Reply:
The method is only used to set the screen resolution in the non-headless session. It should be safe to keep calling it:

```python
            if options.runres and anaconda.gui_mode and not flags.use_rd:
                def on_mutter_ready(observer):
                    set_resolution(options.runres)
                    observer.disconnect()
```

Reply:
ah, I missed the `not` before `flags.use_rd`.

Reply:
I wonder if it'd not be better to generate an monitors.xml file instead, as that'd avoid mutter first setting one resolution, then changing to the one here. It would require copying the way mutter decides how to index port numbers (e.g. DP-#).

Reply:
That's just libdisplay-info, no? IIRC, most Wayland compositors use that to enumerate outputs.

Reply:
No, the enumeration comes from the KMS uAPI. libdisplay-info is to get info about the display e.g. colorimetry information and stuff like that. In mutter the connector name comes from here: https://gitlab.gnome.org/GNOME/mutter/-/blob/e4dd50e1dd7e6a741eb645ec5526768e33082259/src/backends/native/meta-kms-connector.c#L1562-1593 which effectively is a prefix (e.g. `DP-` or `HDMI-` and the an ID from KMS). That isn't really API though, maybe it'll one day try to be smarter.

Reply:
BTW, if I drop the `--virtual-monitor` option in the headless mode, it does not seem to be happy at all:
![no_virtual_monitor](https://github.com/rhinstaller/anaconda/assets/829558/44115c03-cd04-4b9e-a85b-4f223cdeb801)

All this before it becomes possible to connect - I guess it default to 0x0 pixels resolution or something ?